pandas/tests/indexes/test_setops.py

```diff
@@ -501,6 +501,81 @@ def check_intersection_commutative(left, right):
     assert idx.intersection(idx_non_unique).is_unique
 
 
+@pytest.mark.parametrize(
+    "cls",
+    [
+        Int64Index,
+        Float64Index,
+        DatetimeIndex,
+        CategoricalIndex,
+        TimedeltaIndex,
+        lambda x: Index(x, dtype=object),
```

Comment:
PeriodIndex, IntervalIndex, UInt64Index?

Reply:
IntervalIndex is falling back to MultiIndex, which is not working, Would like to tackle this as a follow up

Added tests for the other cases